### PR TITLE
stats: Fix 32-bit limit for byte offsets in stats set, and guard against very high max-stats that will cause shared memory to explode.

### PR DESCRIPTION
**Pull request by jmarantz**
*Description*:
>Allows >4G of stats memory to be addressable by BlockMemoryHashSet by changing uint32_t to size_t for byte-offsets and sizes. Also adds an options-parsing sanity check to avoid huge numbers of stats, which will be successfully allocated, but SIGBUS when being initialized if they exceed the physical capacity of the machine.

*Risk Level*: Low
*Testing*: /test/...
Also tested that 30M stats:
```
bazel-bin/source/exe/envoy-static --config-path  $CFG --mode init_only --max-stats 30000000
```
works now, which SEGV'd prior to this change due to the 32-bit byte-offset limitation.

*Docs Changes*:  updating --max-stats doc in `docs/root/operations/cli.rst`
*Release Notes*: N/A

Fixes https://github.com/envoyproxy/envoy/issues/3463

### CODE REVIEW

File: source/common/common/block_memory_hash_set.h [hunk 1]
```diff
@@ -292,7 +292,7 @@ template <class Value> class BlockMemoryHashSet : public Logger::Loggable<Logger
   struct Control {
     BlockMemoryHashSetOptions options; // Options established at map construction time.
     uint64_t hash_signature;           // Hash of a constant signature string.
-    uint32_t num_bytes;                // Bytes allocated on behalf of the map.
+    size_t num_bytes;                  // Bytes allocated on behalf of the map.
```
Comment (mattklein123) [inline source/common/common/block_memory_hash_set.h]: nit: can we use uint64_t here and elsewhere to be super clear about the sizes?

File: source/server/options_impl.cc [hunk 2]
```diff
@@ -123,13 +123,20 @@ OptionsImpl::OptionsImpl(int argc, const char* const* argv,
...
-        "error: the 'max-obj-name-len' value specified ({}) is less than the minimum value of 60",
-        max_obj_name_len.getValue());
-    std::cerr << message << std::endl;
-    throw MalformedArgvException(message);
-  }
+  auto check_numeric_arg = [](bool is_error, uint64_t value, absl::string_view pattern) {
+    if (is_error) {
+      const std::string message = fmt::format(std::string(pattern), value);
+      std::cerr << message << std::endl;
+      throw MalformedArgvException(message);
```
Comment (mattklein123) [inline source/server/options_impl.cc]: Can we add some options_impl_tests for this?

File: source/common/common/block_memory_hash_set.h (hunk 1 above)
Comment (jmarantz) [reply source/common/common/block_memory_hash_set.h]: ok; done

File: source/server/options_impl.cc (hunk 2 above)
Comment (jmarantz) [reply source/server/options_impl.cc]: There already was a test for --max-obj-name, added one for --max-stats.

File: test/server/options_impl_test.cc [hunk 3]
```diff
@@ -159,4 +159,12 @@ TEST(OptionsImplTest, BadObjNameLenOption) {
     EXPECT_THAT(e.what(), HasSubstr("'max-obj-name-len' value specified"));
   }
 }
+TEST(OptionsImplTest, BadMaxStatsOption) {
+  try {
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply test/server/options_impl_test.cc]: Very nice; there were a few places in that file that needed that transform, so I just did them all.

File: source/server/hot_restart_impl.cc [hunk 4]
```diff
@@ -37,7 +37,7 @@ static BlockMemoryHashSetOptions blockMemHashOptions(uint64_t max_stats) {
   return hash_set_options;
 }
 
-SharedMemory& SharedMemory::initialize(uint32_t stats_set_size, Options& options) {
+SharedMemory& SharedMemory::initialize(size_t stats_set_size, Options& options) {
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/server/hot_restart_impl.cc]: done; SharedMemory::configure also uses a size_t args; should I change all of them?

I feel like there's an argument that size_t is more correct for many of these uses; if we hypothetically built on a 32-bit system, we might not want sizes to be 64 bits.  I'm not sure if we'd ever want Envoy running on some IoT thing someday (or if IoT things are 32 bits). But for the moment this seems fine.

File: source/server/hot_restart_impl.h [hunk 5]
```diff
@@ -48,7 +48,7 @@ class SharedMemory {
    * Initialize the shared memory segment, depending on whether we should be the first running
    * envoy, or a host restarted envoy process.
    */
-  static SharedMemory& initialize(uint32_t stats_set_size, Options& options);
+  static SharedMemory& initialize(size_t stats_set_size, Options& options);
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/server/hot_restart_impl.h]: done

File: source/server/hot_restart_impl.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/server/hot_restart_impl.cc]: I think given that we had overflow/size issues here, my preference would be to be explicit about all sizes and covert the rest of the size_t. We can worry about 32-bit build later if it ever comes to that. If you are ok with that, please go for it.

[the remark this answers is not in the document]
Comment (jmarantz) [reply source/server/hot_restart_impl.cc]: alright; I did kind of a one to two level search throughout the stats system and converted many (but not all) size_t to uint64_t. 

I did not convert size_t usage as indexes and lengths of strings (e.g. `IntervalSet<size_t>`) that were not in the flow of the overall stats-allocation.

**Comment by mattklein123** [comment]
@jmarantz OSX failure looks legit

**Comment by jmarantz** [comment]
Declaration error fixed; sorry about that; OSX passes now.

**Pull request merged**